src/collapse/xctrace.rs

```diff
@@ -0,0 +1,455 @@
+        x.into_owned().into_bytes()
+    } else {
+        x.into_owned()
+    }
+}
+
+fn get_ref_id_from_attributes(attributes: &Attributes) -> io::Result<u64> {
+    let ref_id = attributes
+        .clone()
+        .filter_map(|x| x.ok())
+        .find_map(|x| (x.key.into_inner() == REF).then_some(x.value));
+    let ref_id = match ref_id {
+        Some(x) => x,
+        None => return invalid_data_error!("No ref id found in attributes"),
+    };
+    let ref_id = String::from_utf8_lossy(&ref_id);
+    match ref_id.parse() {
+        Ok(x) => Ok(x),
+        Err(e) => invalid_data_error!("Unrecognized ref id: {}: {:?}", ref_id, e),
+    }
+}
+
+fn get_id_from_attributes(attributes: &Attributes) -> io::Result<u64> {
+    let id = attributes
+        .clone()
+        .filter_map(|x| x.ok())
+        .find_map(|x| (x.key.into_inner() == ID).then_some(x.value));
+    let id = match id {
+        Some(x) => x,
+        None => return invalid_data_error!("No id found in attributes"),
+    };
+    let id = String::from_utf8_lossy(&id);
+    match id.parse() {
+        Ok(x) => Ok(x),
+        Err(e) => invalid_data_error!("Unrecognized id: {}: {:?}", id, e),
+    }
+}
+
+fn get_name_from_attributes(attributes: &Attributes) -> io::Result<Vec<u8>> {
+    let name = attributes
+        .clone()
+        .filter_map(|x| x.ok())
+        .find_map(|x| (x.key.into_inner() == NAME).then_some(x.value));
+    match name {
+        Some(x) => Ok(unescape_xctrace_text(x)),
+        None => invalid_data_error!("No name(symbol) found in attributes"),
+    }
+}
+
+/// Extract necessary info from attributes for constructing backtrace.
+fn attributes_to_backtrace(attributes: &Attributes) -> io::Result<u64> {
+    get_id_from_attributes(attributes)
+}
+
+/// Extract necessary info from attributes for constructing frame.
+fn attributes_to_frame(attributes: &Attributes) -> io::Result<(u64, Vec<u8>)> {
+    let id = get_id_from_attributes(attributes)?;
+    let name = get_name_from_attributes(attributes)?;
+    Ok((id, name))
+}
+
+/// A stack collapser for the output of `xctrace export`.
+#[derive(Default)]
+pub struct Folder;
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, reader: R, writer: W) -> std::io::Result<()>
+    where
+        R: std::io::BufRead,
+        W: std::io::Write,
+    {
+        let mut reader = Reader::from_reader(reader);
+        self.collapse_inner(&mut reader, writer)
+    }
+
+    fn is_applicable(&mut self, input: &str) -> Option<bool> {
+        let mut input = input.as_bytes();
+        let mut line = String::new();
+        loop {
+            if let Ok(n) = input.read_line(&mut line) {
+                if n == 0 {
+                    break;
+                }
+            } else {
+                return Some(false);
+            }
+
+            let trimmed = line.trim();
+            if !trimmed.is_empty() {
+                return Some(trimmed.contains(r#"<?xml version="1.0"?>"#));
```

Comment:
Is there something more distinct we can look for than just whether it's an XML file?